# Main entrypoint of the workflow.
# Please follow the best practices:
# https://snakemake.readthedocs.io/en/stable/snakefiles/best_practices.html,
# in particular regarding the standardized folder structure mentioned there.

import pandas as pd
import yaml

# Import configuration files
params = yaml.load(open("config/params.yml", "r"), Loader=yaml.SafeLoader)
features = yaml.load(open("config/features.yml", "r"), Loader=yaml.SafeLoader)
samples = pd.read_table("config/samples.tsv", comment="#", dtype="str")

# Generate useful variables
SAMPLES = samples["sample_id"].drop_duplicates().values.tolist()
SAMPLE_LIBRARY = samples[["sample_id", "library_id"]].values.tolist()
BAM_REPORTS = ["stats.tsv", "flagstats.txt", "idxstats.tsv"]
KRAKEN2_DBS = [] if features["kraken2_databases"] is None else list(features["kraken2_databases"].keys())
MAG_CATALOGUES = features["mag_catalogues"]
COVERM_GENOME_METHODS = params["quantification"]["coverm"]["genome"]["methods"]
COVERM_CONTIG_METHODS = params["quantification"]["coverm"]["contig"]["methods"]
HOST_NAMES = [] if features["hosts"] is None else list(features["hosts"].keys())
LAST_HOST = HOST_NAMES[-1] if len(HOST_NAMES) > 0 else None

# Import subworkflows
include: "rules/folders.smk"
include: "rules/helpers/_main.smk"
include: "rules/reads/_main.smk"
include: "rules/reference/_main.smk"
include: "rules/preprocessing/_main.smk"
include: "rules/quantification/_main.smk"
include: "rules/report/_main.smk"


rule all:
    """Default pipeline"""
    input:
        # rules.reads.input,
        # rules.reference.input,
        # rules.fastp.input,
        rules.kraken2.input,
        # rules.star.input,
        # rules.bowtie2.input,
        rules.report_step.input,
        rules.report_library.input,
        rules.coverm.input,
